juju/juju
# Dev persist extra networking info to state

## Description of change

The machiner worker collects network device information and invokes `SetObservedNetworkConfig`. This endpoint merges the machiner's config with provider-specific information that includes Network/SubnetID if the provider supports it.

After the config info is fused together, the code makes a call to `SetDevicesAddresses`. However, up till now that API ignored the above bits of information when updating the ip address state document.

This PR threads the additional network-related information through the various API calls and persists the subnetID into the ip address mongo doc.

## QA steps

The following steps test the changes on amazon. In order to properly set up your AWS account for these tests (and avoid unnecessary frustration because your units have no egress) please follow these steps:

1) Go to your account's VPC/subnets [panel](https://console.aws.amazon.com/vpc/home?region=us-east-1#subnets:sort=AvailabilityZone)
2) Create a new subnet with name **isolated**, CIDR **172.31.254.0/24** and AZ **us-east-1a**.
3) Check the existing _public_ subnets and copy the ID of the subnet in AZ **us-east-1a** to your clipboard.
4) Go to the NAT gateway [panel](https://console.aws.amazon.com/vpc/home?region=us-east-1#NatGateways:sort=natGatewayId) and add a new gateway (you might need to allocate a new elastic IP). Paste the **public** subnet ID from step 3 to the subnet text-box.  ⚠️ if you assign the private subnet at this step egress traffic will not work and you will have to start over.
5) Go to the Route table [panel](https://console.aws.amazon.com/vpc/home?region=us-east-1#RouteTables:sort=routeTableId) and create a new route table. Then click on the route table entry, navigate to the **subnet associations** tab and edit the associations list to include the **isolated** subnet. Finally, click on the **Routes** tab, edit the routes and add a **0.0.0.0/0** with the NAT gateway as the target.

```console
$ juju bootstrap aws test --credential $your_aws_cred
$ juju reload-spaces
$ juju add-space isolated 172.31.254.0/24
$ juju deploy mysql --constraints spaces=isolated
```

If you monitor juju status you will see a couple of failed attempts to find the requested subnet. However, eventually, a machine will spin up and the unit should become available.

Connect to the mongo db and run the following query:

```js
$ db.ip.addresses.find({"subnet-cidr":"172.31.254.0/24"}, {"provider-subnet-id": 1}).pretty()
{
        "_id" : "4c43f0be-50be-41b5-8ee5-579f0116c227:m#0#d#ens5#ip#172.31.254.38",
        "provider-subnet-id" : "subnet-083414ae2690ccc65"
}
```

In the response you should see a populated subnet-id value.

PS: don't forget to clean up the NAT and allocated elastic IPs after completing the QA steps.
